### PR TITLE
Introduce `input_type` to specify input pipeline and re-organize the

### PR DESCRIPTION
**Pull request by stellaraccident**
input phase to be self contained.

* This includes a large amount of fiddling with test flags now that we
default to no input legalization.
* Eliminates all of the v0 shape propagation logic, with the exception
of some expansions happening at func boundaries (which should be
reworked but isn't invasive like the rest of it is).
* Creates a dedicated MHLO integrations/tensorflow input pipeline which
encapsulates some of the weird that goes beyond just MHLO ops (i.e.
deals with things that come from an XLA input, which the core compiler
does not touch).
* Fully legalizes both TF and XLA at the import level, allowing the core
compiler to be invoked with no input pipeline.
* Introduces hard checks on the core compiler boundary to emit useful
errors when input dialects have not been previously legalized away.
* Switches to the CHLO direct broadcast lowerings, which allow us to
directly handle the vast majority of dynamic shapes without involving
the shape/shapex dialect.

### CODE REVIEW

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 1]
```diff
@@ -62,96 +62,116 @@ namespace IREE {
...
+
+  // We also don't handle calls well on the old codepath; until we remove the
+  // use of the CFG we can continue inlining.
+  passManager.addPass(mlir::createInlinerPass());
+
   passManager.addNestedPass<FuncOp>(
       IREE::Flow::createHLOToHLOPreprocessingPass());
 
+  passManager.addNestedPass<FuncOp>(
+      mlir::iree_compiler::IREE::Flow::createHLOToHLOPreprocessingPass());
```
Comment (rsuderman) [inline iree/compiler/Dialect/Flow/Transforms/Passes.cpp:85]: Duplicate of above.

**Review by rsuderman — COMMENTED** [review]
Feels like a good change of layering to me, not very many strong feelings overall. I am curious whether we can remove mhlo entirely from IREE ingestion through upstreaming to MHLO, though I believe we lower some MHLO operations to ops in the FLOW dialect.

**Comment by stellaraccident** [comment]
> Feels like a good change of layering to me, not very many strong feelings overall. I am curious whether we can remove mhlo entirely from IREE ingestion through upstreaming to MHLO, though I believe we lower some MHLO operations to ops in the FLOW dialect.

That would be nice but may be a bit too much like breaking into prison for the near future: there are quite a few things we do because we can based on having a strong target and memory model -- and none of that is defined in MHLO. Consensus on those points would be really costly. In practice, we're probably going to find more need in the future to take some more optimal opinions on things and will always benefit from some divergence.

**Review by benvanik — APPROVED** [review]

File: bindings/tflite/testdata/BUILD [hunk 2]
```diff
@@ -18,6 +18,7 @@ iree_bytecode_module(
     src = "add_dynamic.mlir",
     c_identifier = "iree_tflite_testdata_add_dynamic",
     flags = [
+        "--iree-input-type=mhlo",
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply bindings/tflite/testdata/BUILD:21]: You would think, but this predated any of that I think and is kind of using MHLO as a test artifact. Yolo.

**Pull request merged**